### PR TITLE
Added tests for Exception handling.

### PR DESCRIPTION
**Pull request by bethesque**
Was wondering this morning as I walked to work (the best time for random thoughts) whether or not the Exception -> StandardError thing had happened. As part of investigating this, I realised that there were no tests for it, so have added some.

Also renamed 'exception' to 'error' where appropriate in comments.

### CODE REVIEW

**Comment by bethesque** [comment]
Travis failed due to the old Timeout.

```
An error occurred in an after hook
    Timeout::Error: execution expired
    occurred at /home/travis/build/seancribbs/webmachine-ruby/lib/webmachine/spec/adapter_lint.rb:29:in `__script__'
```

I do not think this is due to any problem with the new tests, but will kick it off again and see.

**Comment by bethesque** [comment]
This is ready to merge. It's only failing on 2.1 which is an unsupported ruby anyway and will be removed after merging https://github.com/webmachine/webmachine-ruby/pull/245

File: lib/webmachine/errors.rb [hunk 1]
```diff
@@ -39,6 +39,6 @@ class InvalidResource < Error; end
 
   # Raised when the client has submitted an invalid request, e.g. in
   # the case where a request header is improperly formed. Raising this
-  # exception will result in a 400 response.
+  # error will result in a 400 response.
```
Comment (r-obert) [inline lib/webmachine/errors.rb]: Even though the exception has the name "Error", it's still an exception. 
I haven't heard "raise an error" much, i think "raise an exception" is more common and understood by all Ruby programmers (afaik).

**Comment by r-obert** [comment]
I don't know the history behind the change of subclassing StandardError instead, but i think it will miss some exceptions that are valid to catch but gets rids of others that probably aren't. `ScriptError` and all its subclasses are valid, the others less so, some very much less so: SystemExit, Fatal, NoMemoryError. 

Still, better to be passed any exception and make a decision on what to do, than to be blind to the exception altogether. I don't use webmachine anymore tho, so not even 2c, only 1c. :smile:

**Comment by bethesque** [comment]
The change from Exception to StandardError happened prior to this, and this is just adding tests to cover it.

As some background, it's not best practice to rescue Exception unless you have some specific reason and really know what you are doing, because it stops you being able to sent INT and KILL signals etc to kill running processes. 

Here's a bit of a more detailed explanation: http://blog.honeybadger.io/ruby-exception-vs-standarderror-whats-the-difference/

**Comment by r-obert** [comment]
yeah.. i know the history of 'rescue Exception', and the arguments against it. 
'StandardError' still isn't enough. While not rescuing Exception solves the problem you mention, there's other exceptions that don't fall into that category. 

`ScriptError` and all its subclasses, as an example:
```
ScriptError
   LoadError
   NotImplementedError
   SyntaxError
```

shouldn't i know if `require` fails? 
shouldn't i know if `SyntaxError` is raised (eg from eval()) ? 
shouldn't i know if userland code raises NotImplementedError, which is not unheard of?

**Comment by r-obert** [comment]
if you even ignore `ScriptError` for a second, user-land exceptions inheriting from StandardError is also by means guaranteed. That includes other libraries you might be using. 

I think the better solution is to ignore what you know you don't want to rescue: SystemExit, NoMemoryError, Interrupt, and let everything else fall through.

**Comment by r-obert** [comment]
i think it's also debatable whether or not NoMemoryError should be ignored, a resource may want to handle that some how, but probably for 99.9% of use-cases ignoring is fine. I guess it's debatable if any of those exceptions should be ignored, only Interrupt has a stronger case than any of the others.

**Comment by r-obert** [comment]
> if you even ignore ScriptError for a second, user-land exceptions inheriting from StandardError is also by no means guaranteed. That includes other libraries you might be using.

Data:
https://github.com/search?l=Ruby&q=%3C+Exception&type=Code&utf8=%E2%9C%93

I doubt those examples fall into the category of disinterest, too.

**Comment by bethesque** [comment]
Thanks for the extra detail @r-obert. You make some good points.

I'm going to merge this PR anyway, as it is only adding test cases for existing behaviour. If anyone would like to go through and change the exceptions that are handled to a more sophisticated subset, then go for it ;)

**Comment by r-obert** [comment]
Sure, no problem.
This is the technique Pry uses for the same problem:
https://github.com/pry/pry/blob/master/lib/pry/exceptions.rb#L6-L24

```ruby
begin
  block.call
rescue Pry::RescuableException => e
end
```

In Webmachine, it could be something like:

```ruby
begin
  resource.run
rescue Webmachine::RescuableException => e
  resource.handle_exception(e)
end
```

I doubt i have the time, but probably not a lot of work.

**Comment by r-obert** [comment]
here's a more complete example, that could probably be adapted without much work:
```ruby
module Rescuey
  def self.===(e)
    case e
    when StandardError then true
    when ScriptError then true
    else false
    end
  end
end

begin
  raise RuntimeError
rescue Rescuey
  puts "rescued"
end

begin
  raise StandardError
rescue Rescuey
  puts "rescued"
end

begin
  require "hfdufhudfhfuihfufgfdhg"
rescue Rescuey
  puts "rescued"
end

begin
  raise Exception
rescue Rescuey
  puts "shouldnt get here"
end
```

**Pull request merged**